src/mon/OSDMonitor.h

```diff
@@ -303,6 +312,151 @@ class OSDMonitor : public PaxosService {
+  void _set_cache_autotuning();
+  int _update_mon_cache_settings();
+
+  struct OSDMemCache : public PriorityCache::PriCache {
+    OSDMonitor *osdmon;
+    int64_t cache_bytes[PriorityCache::Priority::LAST+1] = {0};
+    int64_t committed_bytes = 0;
+    double cache_ratio = 0;
+    uint64_t used_bytes = 0;
+
+    OSDMemCache(OSDMonitor *m) : osdmon(m) {};
+
+    virtual uint64_t _get_used_bytes() const = 0;
+    virtual void _update_used_bytes(size_t size, bool recalc) = 0;
+
+    virtual int64_t request_cache_bytes(
+        PriorityCache::Priority pri, uint64_t total_cache) const {
+      int64_t assigned = get_cache_bytes(pri);
+
+      switch (pri) {
+      // All cache items are currently set to have PRI1 priority
+      case PriorityCache::Priority::PRI1:
+        {
+          int64_t request = _get_used_bytes();
+          return (request > assigned) ? request - assigned : 0;
+        }
+      default:
+        break;
+      }
+      return -EOPNOTSUPP;
+    }
+
+    virtual int64_t get_cache_bytes(PriorityCache::Priority pri) const {
+        return cache_bytes[pri];
+    }
+
+    virtual int64_t get_cache_bytes() const {
+      int64_t total = 0;
+
+      for (int i = 0; i < PriorityCache::Priority::LAST + 1; i++) {
+        PriorityCache::Priority pri = static_cast<PriorityCache::Priority>(i);
+        total += get_cache_bytes(pri);
+      }
+      return total;
+    }
+
+      virtual void set_cache_bytes(PriorityCache::Priority pri, int64_t bytes) {
+        cache_bytes[pri] = bytes;
+      }
+      virtual void add_cache_bytes(PriorityCache::Priority pri, int64_t bytes) {
+        cache_bytes[pri] += bytes;
+      }
+      virtual int64_t commit_cache_size(uint64_t total_cache) {
+        committed_bytes = PriorityCache::get_chunk(
+            get_cache_bytes(), total_cache);
+        return committed_bytes;
+      }
+      virtual int64_t get_committed_size() const {
+        return committed_bytes;
+      }
+      virtual double get_cache_ratio() const {
+        return cache_ratio;
+      }
+      virtual void set_cache_ratio(double ratio) {
+        cache_ratio = ratio;
+      }
+      virtual string get_cache_name() const = 0;
+  };
+
+  struct IncCache : public OSDMemCache {
+    IncCache(OSDMonitor *m) : OSDMemCache(m) {};
+
+    virtual void _update_used_bytes(size_t size, bool recalc) {
+      if (recalc) {
+        used_bytes = osdmon->calculate_used_cache_bytes(false);
+        return;
+      }
+      used_bytes += size;
+    }
+
+    virtual uint64_t _get_used_bytes() const {
+      return used_bytes;
+    }
+
+    virtual string get_cache_name() const {
+      return "OSDMap Inc Cache";
+    }
+
+    uint64_t _get_num_osdmaps() const {
+      return osdmon->inc_osd_cache.get_size();
+    }
+
+    double get_avg_bytes_per_osdmap() const {
+      return (double)_get_used_bytes() / (double)_get_num_osdmaps();
+    }
+  };
+  std::shared_ptr<IncCache> inc_cache;
```

Comment:
could be `std::unique_ptr<>`

Reply:
But then I would have to do std::move when I insert the cache to the priority cache manager since it takes a shared_ptr. See declaration below,
`void insert(const std::string& name, const std::shared_ptr<PriCache> c);`

If conversion is acceptable then I can change it. Please let me know.

Reply:
@sseshasa i see. i am afraid we cannot move the pointer away from probably we could change this part in a follow-up PR.

@markhpc can we update the `PriorityCache::Manager`, so the manager only hold a weak pointer, i.e., the plain pointer instead of a smart pointer? or the other way around let it hold unique_ptr of the managed caches. as IMHO, it's either the manager and the owner of the manager who owns the managed cache, not both of them.

Reply:
@tchaikov @markhpc We can probably pick this up in a follow-up PR.